### DISCUSSION TITLE
MSEgui small examples

### DISCUSSION THREAD

**Discussion by rchastain** [mse-org/mseide-msegui#68] [Show and tell]
Hello! I added new projects to the _examples_ section of [MSEgui Studies](https://gitlab.com/rchastain/msegui-studies).

Please share your little demo projects, and let us gather them somewhere!

**Comment by fredvs** [comment]
Excellent!

Maybe add it to mseuniverse samples?

**Reply by rchastain** [reply]
Why not, if it can encourage others to contribute too.

**Comment by lainz** [comment]
Hi, please enter the UI contest with your existing projects
https://forum.lazarus.freepascal.org/index.php/topic,66221.msg506415.html#msg506415

**Reply by rchastain** [reply]
Very nice things can be done indeed with MSEgui, but I am not very good for that. I will see what I can do.

By the way, I have the source of a very nice music player ([XelPlayer](https://zenway.ru/page/xelplayer)). I quickly retouched it to compile it with latest MSEgui version. It is in Russian and I am not sure that it works. Maybe I will upload it somewhere.

**Reply by rchastain** [reply]
When I have time I will try to translate it in English and to make it work.

**Reply by fredvs** [reply]
Wow, it dates from 2013 and are you able to compile it with last fpc and last msegui?

Very nice program and if you can make it work, big WoW to you, XelPlayer dev and MSEgui!

**Reply by fredvs** [reply]
Hello Roland-Champollion.

I think you've found MSEgui's "Rosetta Stone"!

I saw the XelPlayer code on your CodeBerg git site and this program plays with all the gems of MSEgui.
I also like how it handles ini files and other things before application.run;
He also has a personalized way of dealing with the languages.

A pure shiny gem that will be an excellent how-to-msegui..

**Reply by rchastain** [reply]
Yes, there are certainly many things to learn from that project.

I will translate the interface, so that we can see how it works and what needs to be repaired. (I don't remember exactly, but maybe I changed into comments some parts of the code that wouldn't compile.)

**Reply by fredvs** [reply]
Yes, it would be great to comment on anything that isn't working.
So we can see the evolution of msegui and what to do to make old source work.

**Comment by fredvs** [comment]
Hello Roland!

Here a little demo how to use a msegui-maskedbitmap with bgrabitmap:
[EDIT] The previous zip was not ok, here the updated:
[maskedbitmap_to_bgrabitmap.zip](https://github.com/mse-org/mseide-msegui/files/14398073/maskedbitmap_to_bgrabitmap.zip)

Fre;D

**Reply by rchastain** [reply]
Interesting. Thank you Fred!

**Comment by fredvs** [comment]
Here the brother of previous demo: bgrabitmap to maskedbitmap:
It uses temp streams.
[bgrabitmap_to_maskedbitmap.zip](https://github.com/mse-org/mseide-msegui/files/14474274/bgrabitmap_to_maskedbitmap.zip)

[EDIT] And here without streams, out-of-the-box way:
[bgrabitmap_to_maskedbitmap2.zip](https://github.com/mse-org/mseide-msegui/files/14523251/bgrabitmap_to_maskedbitmap2.zip)

**Reply by rchastain** [reply]
Thank you Fred.

**Comment by fredvs** [comment]
Hello.
Demo of a BGRA-Button.
[EDITED x4] Here with cleaned source:
[bgrabitmap_button.zip](https://github.com/mse-org/mseide-msegui/files/14524429/bgrabitmap_button.zip)


![bgra-msegui_button](https://github.com/mse-org/mseide-msegui/assets/3421249/a54c57e2-9a2c-406a-af3d-afa2575456f8)

**Reply by fredvs** [reply]
Huh, in previous post, there is a [EDITED x3] zip that uses fptimer vs sleep for the face after execute.

**Reply by fredvs** [reply]
... and (I hope the last) there is a [EDITED x4] zip with face.image.options set to _[bmo_masked,bmo_graymask]_ (even nicer!).

**Comment by fredvs** [comment]
Hello.

And here a BGRAButton-hover (Thanks Roland for the tip):

[bgrabitmap_button_hover.zip](https://github.com/mse-org/mseide-msegui/files/14530949/bgrabitmap_button_hover.zip)

![bgra-msegui-button-hoover3](https://github.com/mse-org/mseide-msegui/assets/3421249/7d0b477f-8c84-4c4d-8faf-59a037d3ed23)

**Reply by fredvs** [reply]
I just discover the **tfacelist** component.
Loooot of things to explore and understand there.
Is there a doc somewhere on how to use it?

**Reply by rchastain** [reply]
Let me a moment to search.

**Reply by fredvs** [reply]
The component has a **list.count**.

In the hoverbutton-mseuniverse demo, the list has 3 items.
After some exploration, the first item determines the frame to use when the button is not focused (in my demo = no frame).
The second item set the frame when "hooving" the button (pass with the mouse on it). (in my demo the green rectangle).
The 3th item is for... I dont know, I did not catch for what is it.

Also I have no idea how is done to assign a action for the items (when no focused, when focus, when defocus,,..).
Maybe it is hard-coded by item, I dont know.

**Reply by rchastain** [reply]
Good job Fred. Martin was really a genius.

I see that [tfacelist](https://msegui.net/doc/fpdoc/msegui/tfacelist.html) is used in many examples (tabwidget, roundbutton, cornermask, hoverframe, imagedarkening, imageblend) but I found no specific documentation. Maybe I should search for [tface](https://msegui.net/doc/fpdoc/msegui/tface.html). I also remember I have read something on focus, defocus...

**Reply by rchastain** [reply]
I am very very far to understand how MSEide works. I never understood what are the interfaces. It's on my to-do list to study it.

**Reply by fredvs** [reply]
>The 3th item is for... I dont know, I did not catch for what is it.

Ok, ok, it is when clicking on the button.
[EDIT] The button-hover demo was updated with the 3th item (the red frame rectangle when click + moving caption).

But still a mystery where those actions are assigned, I have to jump into Martin code (asap).

**Reply by rchastain** [reply]
> Ok, ok, it is when clicking on the button.

Yes. Explanation by Martin on the German Lazarus forum:

```
In der Imagelist sind die Ecken/Kanten im Gegenuhrzeigersinn angeordnet. Es sind 3*8 bitmaps für die drei Zustände "normal", "hoover" und "gedrückt". Die bitmaps 6, 14 und 22 müssen ersetzt werden.
```
> But still a mystery where those actions are assigned, I have to jump into Martin code (asap).

Maybe the beginning of an answer:

```
Die MSEgui Haupt-Zeichenfunktion ist

 twidget = class(tactcomponent,iscrollframe,iface,iassistiveclient)
[...]
   procedure paint(const canvas: tcanvas); virtual;

Normalerweise wird twidget.paint() für das top-level-widget von twindow.internalupdate() aufgerufen.
```

**Reply by fredvs** [reply]
>für die drei Zustände "normal", "hoover" und "gedrückt".

Ok, so my nose is still ok and so this is the "hardcoded" action for each of the 3 possible items of the list.

>Normalerweise wird twidget.paint() für das top-level-widget von twindow.internalupdate() aufgerufen.

This one, even with my friend GoogleTranslate is nebulous.

**Reply by rchastain** [reply]
So your problem now (if I understand well) is to fill a imagelist at runtime, instead of using the editor. The solution can probably be found in the code of the editor itself.

But I still believe that it has not much interest and doesn't deserve the pain. A complete demo of an MSEgui app with custom buttons, using _images created previously by another application,_ would be a better idea IMHO. We could make a tutorial.

You see what I mean? An application creating the images, and a demo application using them.

> This one, even with my friend GoogleTranslate is nebulous.

It is not absolutely clear for me neither. Let's continue to investigate. ;)

**Reply by fredvs** [reply]
>So your problem now (if I understand well) is to fill a imagelist at runtime, instead of using the editor. The solution can probably be found in the code of the editor itself.

No there is no more problems now, filling a _imagelist_ at runtime is easy and out of the box. https://github.com/mse-org/mseide-msegui/discussions/68#discussioncomment-8659403 and https://github.com/mse-org/mseide-msegui/discussions/68#discussioncomment-8712546

To fill the imagelist with each draw of aBGRABitmap, you may do this:
[EDIT] Demo of this here: https://github.com/mse-org/mseide-msegui/discussions/68#discussioncomment-8725987


```
var
 amaskedbitmap : tmaskedbitmap;
 aBGRABitmap: TBGRABitmap;
...
   timagelist1.clear;  // the imagelist used for the corner/side
   timagelist1.count := 0 ;
   timagelist1.width := bitmapsize;
   timagelist1.height := bitmapsize; 

// With aBGRABitmap do each image of the 8 corner-side and add it opposite clock turn.

   amaskedbitmap.assign(aBGRABitmap.bitmap); // convert aBGRABitmap to maskedbitmap
   timagelist1.addimage(amaskedbitmap); // add the first image in imagelist.

// do the same for the 7 other images to have timagelist1.count = 8.
```

But now that appears **tfacelist** and that I understood the goal of the list, many, many customization is possible, for each action done on the button.

>But I still believe that it has not much interest and doesn't deserve the pain. A complete demo of an MSEgui app with custom buttons, using images created previously by another application, would be a better idea IMHO. We could make a tutorial.

Yes, I understand your point of view.  But, mainly for the frame-images list that needs 8 images that you have to create with Gimp or with LazPaint is a boring work (for me).

It is way easier to draw by code the 8 corner-side using BGRAbitmap and way easier to customize the frames at runtime.
But, I agree, this is for people who want to give many layout-skins for their applications.
Usual applications dont need this and if only one imagelist is used, maybe it is easier to use Gimp or Lazpaint.
My 0,001 cent of course.

**Reply by fredvs** [reply]
>You see what I mean? An application creating the images, and a demo application using them.

Yes, good idea so we have 2 nice ways, with external design program and by code using BGRABItmap.
This is one of my favorite problem: l'embarras du choix.

**Reply by rchastain** [reply]
Thank you for the example of timagelist loading.

**Comment by fredvs** [comment]
> I am very very far to understand how MSEide works. I never understood what are the interfaces. It's on my to-do list to study it.

Same for me.

**Comment by fredvs** [comment]
Hello.

And here a BGRAButton-hoover (face done by BGRABitmap)  with a BGRAFrame (imagelist of frame done by BGRABitmap for green and red rounded-rectangle):

[bgrabitmap_button_hoover_frame.zip](https://github.com/mse-org/mseide-msegui/files/14544518/bgrabitmap_button_hoover_frame.zip)

![bgra_button_frame](https://github.com/mse-org/mseide-msegui/assets/3421249/b389235e-e283-49c2-a1d1-0be44254fe6d)

**Comment by lainz** [comment]
Awesome work, keep going =)

**Reply by fredvs** [reply]
Thanks!

Here the last touch of BGRAButton: https://github.com/mse-org/mseide-msegui/discussions/78#discussioncomment-8731307
